### PR TITLE
DSL enhancements

### PR DESCRIPTION
**Pull request by FabioLuporini**
Reopening from #642 

- '(u + v.dx).dy' and things like that possible
- staggered now is defined as `staggered = x` for example, not more `(0,1,0)` and `None` corresponds to cartesian grids
- FD shortcuts are now created on demand
- Some misc

Most credits to @mloubout

### CODE REVIEW

File: devito/cgen_utils.py [hunk 1]
```diff
@@ -169,11 +173,17 @@ def _print_ListInitializer(self, expr):
         return "{%s}" % ', '.join([self._print(i) for i in expr.params])
 
     def _print_IntDiv(self, expr):
-        return str(expr)
+        return expr.__str__()
```
Comment (FabioLuporini) [inline devito/cgen_utils.py]: droppable?

File: devito/function.py [hunk 2]
```diff
@@ -1479,7 +1406,7 @@ def _interpolation_indices(self, variables, offset=0):
         for i, idx in enumerate(index_matrix):
             ind_subs = dict([(dim, ind) for dim, ind in zip(self.grid.dimensions, idx)])
             v_subs = [(v, v.subs(ind_subs))
-                      for v in variables if not v.base.function.is_SparseFunction]
+                      for v in variables if not isinstance(v, SparseFunction)]
```
Comment (FabioLuporini) [inline devito/function.py]: why was this necessary?

File: devito/function.py [hunk 3]
```diff
@@ -1655,6 +1580,10 @@ class SparseTimeFunction(AbstractSparseTimeFunction, SparseFunction):
 
     is_SparseTimeFunction = True
 
+    @property
+    def fd_indices(self):
```
Comment (FabioLuporini) [inline devito/function.py]: we can drop this, unused

File: devito/function.py (hunk 2 above)
Comment (mloubout) [reply devito/function.py]: Because `interp/inject` was using `indexify` to make it "easier", now done on functions.

File: devito/cgen_utils.py (hunk 1 above)
Comment (mloubout) [reply devito/cgen_utils.py]: yep

**Comment by mloubout** [comment]
TTI is messed up, produces really bad results.

**Comment by mloubout** [comment]
Toook a while, but here it is.
